static/js/code-highlighter.js

```diff
@@ -9,9 +9,146 @@
+  const firstSourceY = scrolling.offsetTop;
+  // The goal is to make sure we're in the area the source line numbers live.
+  const lineForSizing = document.querySelector('.line-number');
+  const sourceLinesX = lineForSizing.offsetLeft + 6;
+  const lineHeight = lineForSizing.offsetHeight;
+
+  const MAX_NESTING = 10;
+
+  let firstStuck = null;
+  let lastStuck = null;
+  const jitter = 6;
+
+  function extractLineNumberFromElem(elem) {
+    if (!elem.classList.contains('line-number')) {
+      return null;
+    }
+
+    return parseInt(elem.dataset.lineNumber, 10);
+  }
+
+  /**
+   * Walk at a fixed offset into the middle of what should be stuck line
+   * numbers.
+   *
+   * If we don't find a line-number, then we expect that to be due to the
+   * transition from stuck elements to partially-scrolled source lines.  It
+   * means the current set of lines are all stuck.
+   *
+   * If we do find a line-number, then we have to look at the actual line
+   * number.  If it's consecutive with the previous line, then it means the
+   * previous line AND this line are both not stuck, and we should return
+   * what we had exclusive of the previous line.
+   *
+   *
+   */
+  function walkLinesAndGetLines() {
+    let offset = 6;
+    let sourceLines = [];
+
+    // Find a line number that can't possibly be stuck.
+    let sanityCheckLineNum =
+      extractLineNumberFromElem(document.elementFromPoint(
+        sourceLinesX, firstSourceY + offset + lineHeight * MAX_NESTING));
+    // if we didn't find a line, try again with a slight jitter because there
+    // might have been a box boundary edge-case.
+    if (!sanityCheckLineNum) {
+      sanityCheckLineNum =
+        extractLineNumberFromElem(document.elementFromPoint(
+          sourceLinesX, jitter + firstSourceY + offset + lineHeight * MAX_NESTING));
+    }
+
+    // If we couldn't find a sanity-checking line number, then either our logic
+    // is broken or the file doesn't have enough lines to necessitate the sticky
+    // logic.  Just bail.
+    if (!sanityCheckLineNum) {
+      return sourceLines;
+    }
+
+    for (let iLine=0; iLine <= MAX_NESTING; iLine++) {
+      let elem = document.elementFromPoint(sourceLinesX, firstSourceY + offset);
+      if (!elem || !elem.classList.contains('line-number')) {
+        break;
+      }
+
+      let lineNum = parseInt(elem.dataset.lineNumber, 10);
+
+      let expectedLineNum = sanityCheckLineNum - MAX_NESTING + iLine;
+      if (lineNum !== expectedLineNum) {
+        // the line-number's parent is the source-line-with-number
+        sourceLines.push(elem.parentElement);
+      } else {
+        break;
+      }
+
+      offset += lineHeight;
+    }
+
+    return sourceLines;
+  }
+
+  let newlyStuckElements = walkLinesAndGetLines();
+
+  let noLongerStuck = new Set(previouslyStuckElements);
+  let lastElem = null;
+  for (let elem of newlyStuckElements) {
+    elem.classList.add('stuck');
+    noLongerStuck.delete(elem);
+    lastElem = elem;
```

Comment:
I'd prefer to just assign `lastElem` outside the loop using a direct index the same way you do `prevLastElem`. Usually doing it inside the loop means it's there for a particular reason, like if you early-exit the loop, but that's not the case here and it confused me.

Reply:
(I was being a lazy typer.  Oh, my kingdom for a language with [-1] slicing without object creation (a la .slice(-1)[0]! ;)) Fixed.